go/consensus/cometbft/beacon/beacon.go

```diff
@@ -101,47 +106,69 @@ func (sc *serviceClient) GetFutureEpoch(ctx context.Context, height int64) (*bea
 
 func (sc *serviceClient) GetEpochBlock(ctx context.Context, epoch beaconAPI.EpochTime) (int64, error) {
 	now, currentBlk := sc.currentEpochBlock()
-	if epoch == now {
-		return currentBlk, nil
-	}
-
-	// The epoch can't be earlier than the initial starting epoch.
 	switch {
+	case epoch == now:
+		return currentBlk, nil
 	case epoch < sc.baseEpoch:
-		return -1, fmt.Errorf("epoch predates base (base: %d requested: %d)", sc.baseEpoch, epoch)
+		return 0, fmt.Errorf("epoch predates base (base: %d requested: %d)", sc.baseEpoch, epoch)
 	case epoch == sc.baseEpoch:
 		return sc.baseBlock, nil
 	}
 
-	// Find historic epoch.
-	//
-	// TODO: This is really really inefficient, and should be optimized,
-	// maybe a cache of the last few epochs, or a binary search.
-	height := consensus.HeightLatest
-	for {
+	// Try the cache first.
+	if cachedHeight, ok := sc.epochCache.Get(epoch); ok {
+		return cachedHeight.(int64), nil
+	}
+
+	lowHeight, err := sc.backend.GetLastRetainedVersion(ctx)
+	if err != nil {
+		return 0, fmt.Errorf("failed to query last retained version: %w", err)
+	}
+
+	blk, err := sc.backend.GetCometBFTBlock(ctx, consensus.HeightLatest)
+	if err != nil {
+		return -1, err
+	}
+	hiHeight := blk.Height
+	// Start with the latest height as it is possible that currentEpochBlock is not the most up to
+	// date in cases where GetEpochBlock is called during epoch transitions.
+	height := hiHeight
+
+	// Find historic epoch with bounded bisection.
+	const maxIterations = 20 // Should be good enough for most use cases.
+	for range maxIterations {
 		q, err := sc.querier.QueryAt(ctx, height)
 		if err != nil {
-			return -1, fmt.Errorf("failed to query epoch: %w", err)
+			return 0, fmt.Errorf("failed to query epoch: %w", err)
 		}
 
-		var pastEpoch beaconAPI.EpochTime
-		pastEpoch, height, err = q.Epoch(ctx)
+		var (
+			pastEpoch   beaconAPI.EpochTime
+			epochHeight int64
+		)
+		pastEpoch, epochHeight, err = q.Epoch(ctx)
 		if err != nil {
-			return -1, fmt.Errorf("failed to query epoch: %w", err)
+			return 0, fmt.Errorf("failed to query epoch: %w", err)
 		}
 
-		if epoch == pastEpoch {
-			return height, nil
+		switch {
+		case epoch == pastEpoch:
+			_ = sc.epochCache.Put(pastEpoch, epochHeight)
+			return epochHeight, nil
+		case epoch < pastEpoch:
+			hiHeight = epochHeight // epochHeight <= height
```

Comment:
These two comments are confusing, as `epochHeight` is always smaller or equal to `height`.

Reply:
That is exactly what the comment says :P that if the epoch is lower, the smaller should be chosen to potentially skip more steps and `epochHeight` is always smaller. I'll amend the comment to say this in words.

Reply:
Ups, I see they are the same :) I would just remove them.